### ISSUE TITLE
Some dishwasher entities sometimes missing

### ISSUE THREAD

**Issue by emufan** [ekutner/home-connect-hass#89]
## Version of the custom_component
0.5.4

## Describe the bug

Not every time, but often, during runtime of my dishwasher, the buttons start/stop and the entities power state, active program (not on the screenshot, but same there), program progress, rest time, energy forecast and water forecast are not available. Every other entity is populated. Always these same entities are missing, if they are missing, so everything there (25% cases of runs) or exactly these are missing (75% of runs).

![image](https://user-images.githubusercontent.com/22775515/182041327-b0cd0728-ae1c-4dff-b278-80d100bc56e8.png)
![image](https://user-images.githubusercontent.com/22775515/182041481-f3d941c4-2a15-443a-a734-44ae5e0718db.png)

In the BSH app, I can see the attributes
![image](https://user-images.githubusercontent.com/22775515/182041504-0062003c-3e7a-4a2f-861a-0f148c0ed936.png)

See logs attached from yesterdays evening run with these symptoms. There is another from earlier noon yesterday, which I have left there as well. Unfortunately I didn't see, if the entities were filled in this run or not. In the evening run, not, see screenshots above.

![image](https://user-images.githubusercontent.com/22775515/182041524-27684e53-c785-42e9-a605-83cb9a806b83.png)

## Debug log

[home-assistant.log](https://github.com/ekutner/home-connect-hass/files/9229559/home-assistant.log)

**Comment by emufan** [comment]
> if you feel you have enough understanding how the API works

Unfortunately, no, I'm sure that I don't have enough understanding. 😪

> (HTTP GET /api/homeappliances/SIEMENS-SN12345-67890/programs/active) but gets a 404 error in response.

Read this as well, but was not sure if this is the ("only") problem, because I saw in the same run the these as well:

```
2022-07-30 23:12:26 DEBUG (MainThread) [home_connect_async.callback_registery] Broadcasting event: BSH.Common.Option.RemainingProgramTime = 1140

or 

2022-07-30 23:01:15 DEBUG (MainThread) [home_connect_async.callback_registery] Broadcasting event: BSH.Common.Option.ProgramProgress = 79
```

But both entities were not populated either.

![image](https://user-images.githubusercontent.com/22775515/182423468-945e16d9-5df6-42f8-98f8-70e603f32840.png)

**Comment by ekutner** [comment]
The entity is coming from the active program, without it the event doesn't have an object to update.
I will report the issue to HC but I wouldn't hold my breath for a fix.

**Comment by emufan** [comment]
Ah. O.k. so the listed entities belong together. Understand.

Find attached another log from such a case from yesterday evening. Perhaps there are further information included. Behavior was the same. But ended with another result ...

```
HTTP Error 429 - Too Many Requests. Sleeping for 78912 seconds and will retry
```

[home-assistant.log](https://github.com/ekutner/home-connect-hass/files/9250353/home-assistant.log)

**Comment by ekutner** [comment]
It is still the same root cause.  The integration expects to have an active program when it receives a progress update notification, if one doesn't exist locally it tries to get it from the server, however that is failing, so it keeps trying with every progress event until it gets blocked because of too many API calls.
I reported this to HC, lets see if they respond and when...

**Comment by emufan** [comment]
Update from here. Is now occurring for my washing machine as well. With even more unavailable entities than at the dishwasher. Sometimes with different error messages, perhaps it narrows down the bug finding (on BSH or your side).

BUT something is suspicious. I double-checked with core integration. Black screenshots = your, white = Core.

1. There were unavailable entities as well. But the remaining time was there, where it was not there in your integration.
2. Some minutes later, Everything was fine in Core automatically
![image](https://user-images.githubusercontent.com/22775515/184076609-f58ea3bd-0118-4465-a741-41b4291d3160.png)
3. At the same time, your integration shows still everything unavailable.
![image](https://user-images.githubusercontent.com/22775515/184076741-a1e8988b-fff1-403e-9db6-1aa3030ad0a1.png)
4. I waited some time scan interval, but it stays this way. Then I reloaded your integration and the information were there as well
![image](https://user-images.githubusercontent.com/22775515/184076898-c467a9dd-6305-4f4d-90c2-1bc85fca4439.png)

Did I wait not long enough (scan interval, but I waited long) or is the Core integration handling the repair of available information different/better than this one?

Find another Log from this washing machine (WMabcd...) from yesterday evening. Please ignore the parallel run from dishwasher (SN123...)
[home-assistant.log](https://github.com/ekutner/home-connect-hass/files/9306696/home-assistant.log)

**Comment by ekutner** [comment]
The built-in integration has a predetermined set of entities, so when the event is received it can just update the progress. This integration is fully dynamic, so it needs to get the available entities from the API. 
I did add a small, and untested, workaround, so at least you will not get rate limited when this happens, so the integration will stop trying to get the active program after 5 failed attempts.

**Comment by emufan** [comment]
Ah, o.k. understood. But do you have a clue, why the Core worked automatically after the data was there and this integration only after integration-reload?

And did you see the new errors like `API call failed with code=409 error=SDK.Error.WrongOperationState` are they related and follow-up-problems because of the same root-cause as wel?

**Comment by ekutner** [comment]
I don't know why it happened and the log seems to be after you've restarted the integration.
409 errors are normal and expected.

**Comment by emufan** [comment]
> 409 errors are normal and expected.
o.k. Didn't know.

> I don't know why it happened and the log seems to be after you've restarted the integration.
No. Restart happened at ~22:27-22:29, after I saw, that Core has values again and this one not. Before the values were noth there, afterwards they were again.

**Comment by ekutner** [comment]
ok, I see. The problem is that the log has mixed data from your two appliances that were active at the same time so it's very difficult to understand what exactly happened, and I don't even know which one is which.
I suggest you wait until this happens again when only one of the appliances is active and then upload the log.

**Comment by emufan** [comment]
> and I don't even know which one is which.

See post from this morning.

> Find another Log from this washing machine (WMabcd...) from yesterday evening. Please ignore the parallel run from dishwasher (SN123...)

> I suggest you wait until this happens again when only one of the appliances is active and then upload the log.

Yes of course. As always not time critical. 😉

**Comment by bkbartk** [comment]
I actually experience a simular issue,
but in this case the same sensor somtimes is of a diffent type.

`
sensor.bosch_wav28kh7nl_68a40e90e01e_laundrycare_washer_option_spinspeed
select.bosch_wav28kh7nl_68a40e90e01e_laundrycare_washer_option_spinspeed
`
The above is the same thing, but sometimes it's  select, and other times it's a sensor. making it hard to integrate this in the interface.

at the moment I have 2 devices running at the same time so the log probably will mess up stuff too.

If this turn out to be a different issue, I will create a new issue.

**Comment by ekutner** [comment]
@bkbartk I don't think it's not the same thing and anyway, it is expected (and desired) that some objects will exist both as a sensor and as some of the settable entities (select/switch/number...). These entities are used for different things and they both exist at the same time although they may be disabled/unavailable when the context changes. 
The Spin Speed program option is exactly such an object. The "select" entity is used to set its value while the "sensor" entity shows the value. Most of the time they will have the same values, but not necessarily all of the time. For example, when a program is active the "select" may get disabled (depending if your appliance allows changing this option using the API while a program is active) but the sensor will still show the current value.

**Comment by CAP-Team** [comment]
Since a couple of days my dishwasher is unavailable in HA. It is available in the official app. My fridge is working just fine in HA.
I tried removing the dishwasher from the app and re-adding it, but it didn't help.

**Comment by ekutner** [comment]
I'm not sure what you mean that you removed it from the app. You probably need to disconnect it from the HA service and then reconnect. This should involve the appliance itself. Verify that the app access the dishwasher when wifi is turned off on the phone. If it can then please enable debug logging and attach a complete log file.

**Comment by CAP-Team** [comment]
I removed it from the official Home Connect app and then re-added it.
It shows up and works correctly from the official app. 
But in Home Connact Alt it shows up as unavailable.

![image](https://user-images.githubusercontent.com/60254660/189532578-87675922-6e2b-4073-9892-db4e5f59118a.png)

There seem to be no diagnostics or errors in the log file

**Comment by ekutner** [comment]
Did you check it is accessible with wifi off on the HC mobile app?

**Comment by CAP-Team** [comment]
Yes it's pingable and I can operate the dishwasher.

![Screenshot_20220911_194343](https://user-images.githubusercontent.com/60254660/189541741-e5f99fff-4e1d-4c16-81ab-4e3b81b55af4.jpg)

**Comment by ekutner** [comment]
Okay, so please upload a debug log. See in the README how to enable debug logging and also add the following to your home_assistant_alt config section:
log_mode: 5

**Comment by CAP-Team** [comment]
Here are the logs.
Nothing about a dishwasher, just the fridge.

[home-assistant.log](https://github.com/ekutner/home-connect-hass/files/9551868/home-assistant.log)

**Comment by ekutner** [comment]
I'm pretty sure the dishwasher is just not being returned by the API. This is a known issue and is solved by re-registering the appliance from the Home Connect service and then registering it back again. To confirm, please set log_mode: 7 restart HA and set the new log.

**Comment by CAP-Team** [comment]
I completely removed and reinstalled the integration and completely reset the dishwasher to factory defaults and then re-added it to the official Home Connect app. It's now back in Home Connect Alt :)

**Comment by ekutner** [comment]
Happy to hear that. It's the appliance reset that solved the issue, no need to reinstall the integration because it doesn't persist a state. As noted, this is a known issue with the Home Connect service, appliances just stop showing through the API until they are de-registered and re-registered again.
